### PR TITLE
Allows to compute the largest eigenvalue magnitude in SquareMatrix and SymmetricMatrix

### PR DESCRIPTION
**Pull request by regislebrun**
Now one can compute an approximation of the largest eigenvalue module using power iterations. It allows to scale the regularization factor of many algorithms in a meaningful way (GaussianProcess, KarhunenLoeveAlgorithm, GeneralLinearModel and so on)

### CODE REVIEW

File: lib/src/Base/Type/openturns/SquareMatrix.hxx [hunk 1]
```diff
@@ -139,6 +139,9 @@ public:
   ComplexCollection computeEigenValues(const Bool keepIntact = true);
   ComplexCollection computeEV(SquareComplexMatrix & vOut,
                               const Bool keepIntact = true);
+  /** Compute an approximation of the largest eigenValue module using power iterations */
+  Scalar computeApproximateLargestEigenValueModule(const UnsignedInteger maximumIterations = ResourceMap::GetAsUnsignedInteger("Matrix-LargestEigenValueIterations"),
+                                                   const Scalar epsilon = ResourceMap::GetAsScalar("Matrix-LargestEigenValueRelativeError")) const;
```
Comment (sofianehaddad) [inline lib/src/Base/Type/openturns/SquareMatrix.hxx:144]: missing `virtual`

File: lib/src/Base/Type/openturns/SymmetricMatrix.hxx [hunk 2]
```diff
@@ -154,6 +154,9 @@ public:
   Point computeEigenValues(const Bool keepIntact = true);
   Point computeEV(SquareMatrix & vOut,
                   const Bool keepIntact = true);
+  /** Compute an approximation of the largest eigenValue module using power iterations */
+  Scalar computeApproximateLargestEigenValueModule(const UnsignedInteger maximumIterations = ResourceMap::GetAsUnsignedInteger("Matrix-LargestEigenValueIterations"),
```
Comment (sofianehaddad) [inline lib/src/Base/Type/openturns/SymmetricMatrix.hxx:158]: missing `override`

File: python/src/HMatrixImplementation_doc.i.in [hunk 3]
```diff
@@ -175,7 +175,16 @@ OT_HMatrix_assembleTensor_doc
...
-    Factorization method, either one of: LDLt, LLt or LU"
+    Factorization method, either one of: LDLt, LLt or LU
+
+Notes
+-----
+The factorization embeds an automatic regularization procedure based
+on an approximation of the largest eigenvalue module. Its computation
+is done using a power iteration, controlled by the
+'HMatrix-LargestEigenValueRelatveError' and
+'HMatrix-LargestEigenValueIterations keys in the :class:`~openturns.ResourceMap`.
```
Comment (sofianehaddad) [inline python/src/HMatrixImplementation_doc.i.in:186]: 'HMatrix-LargestEigenValueIterations' --> missing '

**Review by sofianehaddad — COMMENTED** [review]
Minor remarks :
 - `10` seems small for `LargestEigenValueIterations`
 - In other experiments, we use lower `relative precision` (`~ 1e-6`). What is the impact if we use `1e-1`?
 - In case `precision` is not reached, do we prefer a `warning` or an `exception` ?  If exception is thrown we might improve precision (in a `kriging` usage for example)

**Review by jschueller — APPROVED** [review]

**Comment by jschueller** [comment]
I suggest a more commit message: "Matrix: Allow to compute largest eigen value"
It's not wired anywere yet, is it suppose to be used in an upcoming PR, or already useful as is ?

File: lib/src/Base/Type/openturns/SquareMatrix.hxx (hunk 1 above)
Comment (regislebrun) [reply lib/src/Base/Type/openturns/SquareMatrix.hxx:144]: Why do you want to make this method virtual? None of the other methods are virtual...

File: lib/src/Base/Type/openturns/SymmetricMatrix.hxx (hunk 2 above)
Comment (regislebrun) [reply lib/src/Base/Type/openturns/SymmetricMatrix.hxx:158]: Could you recall me the meaning of the override keyword? I don't see why it should be put for this specific method and not for e.g. computeEV().

File: python/src/HMatrixImplementation_doc.i.in (hunk 3 above)
Comment (regislebrun) [reply python/src/HMatrixImplementation_doc.i.in:186]: Here, `ResourceMap.GetAsUnsignedInteger("HMatrix-LargestEigenValueIterations")` returns 10. How do you get this error?

Comment (sofianehaddad) [reply python/src/HMatrixImplementation_doc.i.in:186]: no it is a typo. There is a missing "'" (after `LargestEigenValueIterations` word)

File: lib/src/Base/Type/openturns/SymmetricMatrix.hxx (hunk 2 above)
Comment (sofianehaddad) [reply lib/src/Base/Type/openturns/SymmetricMatrix.hxx:158]: `SymmetricMatrix` inherits from `SquareMatrix` and parent class has its own `computeApproximateLargestEigenValueModule` method. Thus you want to override the parent method and use this one for `SymmetricMatrix`

**Comment by regislebrun** [comment]
The purpose of this method is to get an approximation of the largest eigenvalue module of a given matrix. In the case of a symmetric matrix, it is the largest absolute value as all the eigenvalues are real, and in the case of covariance matrices it is the largest eigenvalue as they are both real and positive. It allows to get the correct order of magnitude for the automatic regularization loops, avoiding many iterations e.g. in GaussianProcess or Kriging. Hence, I deliberately limited the number of iterations and the relative precision. But now that it is exposed as a method of SquareMatrix and SymmetricMatrix, I am not sure that this choice is reasonable. Maybe we should set more stringent values as a default, and adapt them in the different contexts.
@jschueller This method is not yet called in the other algorithms because it is intended to be used in the CovarianceModel::DiscretizeAndFactorize() method, and Sofiane asked me to expose it in a dedicated PR. I will use it directly in my work on Kriging, and he will use it in his work on covariance models.

File: lib/src/Base/Type/openturns/SquareMatrix.hxx (hunk 1 above)
Comment (sofianehaddad) [reply lib/src/Base/Type/openturns/SquareMatrix.hxx:144]: Because you need to change method content for children classes

**Comment by regislebrun** [comment]
So what is your opinion concerning the default values for the number of iterations and the relative precision? The algorithm converges linearly (we gain a digit at a constant number of iteration) and its convergence depends on the spectral gap between the largest eigenvalue in module r_1 and the second largest one r_2. The error is reduced by a factor of r_2/r1 at each iteration.
Another question is the order of the arguments for this method: iterations first, followed by precision, or the converse?

**Comment by regislebrun** [comment]
BTW, the name could be shorten by removing 'approximate' if the default parameters lead to 6 significant digits...

File: lib/src/Base/Type/openturns/SymmetricMatrix.hxx (hunk 2 above)
Comment (regislebrun) [reply lib/src/Base/Type/openturns/SymmetricMatrix.hxx:158]: Ok, I overlooked this point. I was shadowing the method instead of overloading it...

**Comment by sofianehaddad** [comment]
@regislebrun First we could effectively short the names and remove the `approximate`.

I am aware of difference between `SquareMatrix` & `SymmetricMatrix` and my purpose was only to overload properly methods.
Concerning values, I think we could set a better precision (let us say for example `1e-4`) and increase iterations number (`50` iterations for example). I do not remember the various observations we made some years ago on iterations needed to reach a specific precision target (for example, `1e-4`, `1e-6`, `1e-8`) depending on `size`, `covariance type`, `mesh`...
Btw, we should make sure algo is nearing the largest eigen value module such as regularization in `GP`, `kriging` makes sense. Also I am still wondering what to do if precision is not reached...

**Comment by regislebrun** [comment]
A few remarks here:
+ In our case (SPD matrices, at least theoretically) the largest eigenvalue module should always be the largest eigenvalue, as the negative ones should be close to zero. 
+ If the expected precision is not reached, this is not an issue for the regularization procedure, as the actual largest module is always greater than the computed one. We should only have to regularize a little bit more.
+ To do that, I propose to implement the same strategy as in HMatrixImplementation::factorize(), it means to take the approximation as is and to loop over the additional multiplicative factor until the factorization succeeds.

An exception does not allow to return a value so it cannot be the only solution. A warning is too silent for other applications. A return code would be perfect but we didn't choose this solution. Maybe I can add a flag to the method, or an option to the LOG class to convert warnings into exceptions?

**Comment by regislebrun** [comment]
Or maybe the method should throw an exception at the SquareMatrix/SymmetricMatrix level, and not at the MatrixImplementation level? Something like:
```
UnsignedInteger computeLargestEigenValueModuleSym(Scalar & module, const UnsignedInteger maximumIterations = ResourceMap::GetAsUnsignedInteger("Matrix-LargestEigenValueIterations"), const Scalar epsilon = ResourceMap::GetAsScalar("Matrix-LargestEigenValueRelativeError") const;
```
where the best approximation is returned through a modified argument and a code is returned. This method is called directly by the factorize() method, and an exception is triggered thanks to this code in SquareMatrix and SymmetricMatrix.

**Comment by sofianehaddad** [comment]
What do you mean by "the actual largest module is always greater than the computed one." ?
Indeed, a `flag` might be the solution.

**Comment by regislebrun** [comment]
I mean: the true value of the largest magnitude of the eigenvalues is greater than the approximation obtained by power iterations.

**Comment by regislebrun** [comment]
Is the solution based on a return code ok for you?

**Comment by jschueller** [comment]
> Is the solution based on a return code ok for you?

should it be a bool instead if it's only true/false ?

**Comment by regislebrun** [comment]
I was wondering if a more detailed status could be returned. Usually one returns a code (e.g. lapack), but I can easily change it: let's go for a bool. Remember that these methods are intended to be used only by developers.

**Review by sofianehaddad — APPROVED** [review]

**Pull request merged**